### PR TITLE
Add a bunch of C# styles

### PR DESCRIPTION
**Pull request by matiasbeckerle**
_Disclaimer: even if you are not added as reviewer, please feel free to participate with an emoticon, comment, approval, idea, PR, issue, insights, arguments, code example, whatever. Join the discussion! @AlphaGit and @mvazquez-ms will be the tie-breakers when we can't found a consensus._

## Background

Now that we have defined a basic layout for adding C# styles, we need to add them. This PR only aims to provide some styles.

## Changes

- Added a bunch of C# styles.

## Notes

- Consider using "view" mode for reviewing because seeing the resulted markdown it's easier.
- I'm adding only a bunch of them so reviews are easier (and is an amount of work I can add too).
- Let me know about grouping & order.
- Maybe next week I can add another round of styles.

### CODE REVIEW

File: CSharp/README.md [hunk 1]
```diff
@@ -23,6 +24,115 @@ var httpClient = new HttpClient();
...
+
+```csharp
+// Bad
+this.ValidateParameters();
+
+// Good
+ValidateParameters();
+```
+
+### Use *string* instead of *String* keyword
```
Comment (andresmoschini) [inline CSharp/README.md]: I agree with this convention, but what do you think about to generalize it for all Language keywords (int, string, etc. ) vs framework type names (Int32, String, etc)?

File: CSharp/README.md [hunk 2]
```diff
@@ -23,6 +24,115 @@ var httpClient = new HttpClient();
...
+
+```csharp
+// Good
+public string Name { get; private set; }
+
+// Good
+public string UppercaseName => Name.toUpperCase(); 
+```
+
+### Don't use braces for just one line
```
Comment (andresmoschini) [inline CSharp/README.md]: Some time ago there was a discussion about it, after research many other conventions we concluded that it was a good idea to always use braces, the reason is mainly related to version control tools, diffs are simpler when you add a line and it is less error prone on resolving conflicts. There were really persuasive examples.

File: CSharp/README.md (hunk 1 above)
Comment (AlphaGit) [inline CSharp/README.md]: +1. In the past we have worded this as "use language built-in alias instead of class names"

File: CSharp/README.md [hunk 3]
```diff
@@ -23,6 +24,115 @@ var httpClient = new HttpClient();
...
+// Bad
+String.IsNullOrEmpty(name);
+
+// Good
+string.IsNullOrEmpty(name);
+```
+
+### Use readonly fields when possible
+
+### Use string interpolation
```
Comment (AlphaGit) [inline CSharp/README.md]: Should we phrase this as "Prefer string interpolation to string.Format"? Otherwise, people would start creating interpolated strings when there are no parameters to interpolate. (But really, it's kind of an edge case, so not very important.)

File: CSharp/README.md [hunk 4]
```diff
@@ -23,6 +24,115 @@ var httpClient = new HttpClient();
...
+public void SomeMethod()
+{
+  DoSomethingFirst();
+  DoSomerthingLater();
+}
+
+// Good
+public void SomeMethod()
+{
+    DoSomethingFirst();
```
Comment (AlphaGit) [inline CSharp/README.md]: The JavaScript code style had a clever way of showing this: it used a particular character to display the spaces.

Do you think it would be a good idea to show that here too?

File: CSharp/README.md [hunk 5]
```diff
@@ -23,6 +24,115 @@ var httpClient = new HttpClient();
...
+## Layout
+
+### Use *four spaces* per indentation level
+
+```csharp
+// Bad
+public void SomeMethod()
+{
+  DoSomethingFirst();
+  DoSomerthingLater();
```
Comment (AlphaGit) [inline CSharp/README.md]: Typo: Some**r**thing.

File: CSharp/README.md [hunk 6]
```diff
@@ -23,6 +24,115 @@ var httpClient = new HttpClient();
...
+}
+
+// Good
+public void SomeMethod()
+{
+    DoSomethingFirst();
+    DoSomerthingLater();
```
Comment (AlphaGit) [inline CSharp/README.md]: Typo: Some**r**thing.

File: CSharp/README.md [hunk 7]
```diff
@@ -23,6 +24,115 @@ var httpClient = new HttpClient();
...
+}
+```
+
+### Remove unused or redudant namespaces
```
Comment (AlphaGit) [inline CSharp/README.md]: Typo: Redu**n**dant.

File: CSharp/README.md [hunk 8]
```diff
@@ -23,6 +24,115 @@ var httpClient = new HttpClient();
...
+
+### Use single line and lambda getters
```
Comment (AlphaGit) [inline CSharp/README.md]: "for simple methods and read-only properties", what do you think about that addition?

File: CSharp/README.md (hunk 2 above)
Comment (AlphaGit) [inline CSharp/README.md]: Do you have some links about that? While I understand the point, it kind of gets in the way of early return patterns where you can simplify a lot the readability of some code by performing simple checks and returning early, something that we also want to encourage.

This is what we have in mind:

```csharp
if (!Model.Valid()) throw new InvalidSomethingException();
if (Model.DoNothing) return;
if (DBModel.WasAlreadyUpdated) return;

// perform update and changes...
```

Maybe, for cases like this, we could encourage the statement to be written in the same line as the condition.

File: CSharp/README.md [hunk 9]
```diff
@@ -23,6 +24,115 @@ var httpClient = new HttpClient();
...
+{
+    return "A payment is required.";
+}
+
+// Good
+if (payment == null)
+    return "A payment is required.";
+```
+
+### Use spaces in array/object initializers
```
Comment (AlphaGit) [inline CSharp/README.md]: For later discussion: opinions about dangling list comma?

File: CSharp/README.md [hunk 10]
```diff
@@ -23,6 +24,115 @@ var httpClient = new HttpClient();
...
+var result = 4/2;
+
+// Bad
+var result = 4/2+6;
+
+// Good
+var result = 4 / 2;
+
+// Good
+var result = (4 / 2) + 6;
```
Comment (AlphaGit) [inline CSharp/README.md]: Also to add: avoid space before opening parenthesis/brackets, avoid space after closing parenthesis/brackets.

File: CSharp/README.md (hunk 1 above)
Comment (mvazquez-ms) [inline CSharp/README.md]: +1 and I would try to do something like this with every rule when possible:

-----
Rule : "Use built-in type alias"
// Good example
// Bad Example
Related .editorconfig settings: `dotnet_style_predefined_type_for_locals_parameters_members = true:warning`
StyleCop Rule: [SA1121](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1121.md) 

-----
I know it's a lot of work, but then we can easily customize automatic static code analysis, by enabling/disabling rules.

(The end result of this guide should be pre-defined .editorconfig, .vssetting and .ruleset files.

File: CSharp/README.md (hunk 2 above)
Comment (andresmoschini) [inline CSharp/README.md]: Well, a quick googling did not me examples about code merge resolution, only more mentions about it hehe.

But I found other reasons.

* ASP.NET team use them, for example [ActionDescriptorExtensions](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Abstractions/Abstractions/ActionDescriptorExtensions.cs#L22), [FilterContext](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/FilterContext.cs#L24), [ModelBindingResult](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs#L59), [MvcServiceCollectionExtensions](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc/MvcServiceCollectionExtensions.cs#L27), the are using early returns and in my opinion it has not readibility issues. Maybe the code is more extense vertically, but I think that it is not a problem.

* It is [less error prone](https://softwareengineering.stackexchange.com/questions/16528/single-statement-if-block-braces-or-no/16530#answer-16530)

    ```csharp
    if(condition) 
    //    statement;
    otherStatement;
    ```

    ```csharp
    if(condition) 
        statement;
        otherStatement;
    ```

**Comment by lmarquezMS** [comment]
Just to know, basic conventions as using PascalCase to define properties, field naming using dashes, etc, will be defined in this document or there are given as known.

File: CSharp/README.md (hunk 1 above)
Comment (AlphaGit) [inline CSharp/README.md]: I think it's a great idea, but for @matiasbeckerle, maybe he wants to tackle one of these at a time. If you don't want to do that now, I think it's ok.

**Comment by AlphaGit** [comment]
@lmarquezMS They will be also defined here... eventually.

File: CSharp/README.md (hunk 1 above)
Comment (gastonmancini) [inline CSharp/README.md]: +1

File: CSharp/README.md (hunk 4 above)
Comment (gastonmancini) [inline CSharp/README.md]: Agree with @AlphaGit.

Maybe something like this code can work...


```csharp
// Bad
public void SomeMethod()
{
 ・DoSomethingFirst();
 ・DoSomethingLater();
}

// Good
public void SomeMethod()
{
 ・・・DoSomethingFirst();
 ・・・DoSomerthingLater();
}
```


I would also add a link to the IDE documentation on how to configure the identation. Many people forgot about this.

I also use the IDE with the "View White Space" option enabled. The keyboard shortcut is CTRL+R, CTRL+W.

File: CSharp/README.md (hunk 1 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: Thank you all. I will do this improvement but not sure about doing the .editorconfig, .vssetting and .ruleset thing now. I do believe we should add them but in the worst scenario we can add those later, once we define the rules.

File: CSharp/README.md (hunk 2 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: I will let @AlphaGit & @mvazquez-ms chose the preference here. Anyways we should remember this is only a recommendation guide.

File: CSharp/README.md (hunk 1 above)
Comment (mvazquez-ms) [inline CSharp/README.md]: > but not sure about doing the .editorconfig, .vssetting and .ruleset thing now

Sure, I said

> The end result of **this guide**

Not this PR. We already have those files defined on BetterCare, so it will be a matter of grabbing them and add a few comments.

File: CSharp/README.md (hunk 3 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: Good idea. Will rephrase.

File: CSharp/README.md (hunk 4 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: I will check it, but yes, for sure we can.

File: CSharp/README.md (hunk 5 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: Time to install a spellchecker on VSCode. Thanks!

File: CSharp/README.md (hunk 6 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: At least I'm consistent 😄

File: CSharp/README.md (hunk 7 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: Thanks, will fix.

File: CSharp/README.md (hunk 8 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: That I like it and I will do it.

File: CSharp/README.md (hunk 9 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: Not sure to follow... example?

File: CSharp/README.md (hunk 10 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: Good idea!

**Comment by matiasbeckerle** [comment]
As @AlphaGit said, yes. We are going to add those styles too @lmarquezMS.

File: CSharp/README.md (hunk 9 above)
Comment (mvazquez-ms) [inline CSharp/README.md]: > Not sure to follow... example?

[Trailing comma - SA1413](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1413.md)

I'm OK if we want to add that to our code style (the diffs looks nicer if we apply that rule)

File: CSharp/README.md (hunk 2 above)
Comment (mvazquez-ms) [inline CSharp/README.md]: I don't have a strong opinion, but I prefer not to force the usage of braces everywhere. VS automatic indentation of the code makes it very unlike to end up with a piece of code looking like:

```
if (condition) 
    statement;
    otherStatement;
```

File: CSharp/README.md (hunk 10 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: Done, but maybe you prefer to paraphrase?

File: CSharp/README.md (hunk 7 above)
Comment (gastonmancini) [inline CSharp/README.md]: I would rephrase this to: "Remove unused or redundant 'using' statements (or directives)"

File: CSharp/README.md (hunk 9 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: I don't particularly like that style 😄 I can add it, if you want to. Curious @mvazquez-ms, how that extra comma helps in diffs?

**Comment by matiasbeckerle** [comment]
Ok, ready for another round of reviews. BTW check the group order if you can. I should create a group for one of these rules? Should I move one to other group? Let me know. Appreciated all the feedback!

File: CSharp/README.md [hunk 11]
```diff
@@ -23,6 +24,118 @@ var httpClient = new HttpClient();
...
+### Use readonly fields when possible
+
+### Prefer string interpolation to *string.Format*
+
+```csharp
+// Bad
+var message = string.Format("My name is {0} {1}.", person.FirstName, person.LastName);
+
+// Good
+var message = $"My name is {person.FirstName} {person.LastName}.");
```
Comment (mvazquez-ms) [inline CSharp/README.md]: You have an extra `)` on this line.

File: CSharp/README.md (hunk 9 above)
Comment (mvazquez-ms) [inline CSharp/README.md]: > how that extra comma helps in diffs?

When you modify a multi-line initializer, you have to add a comma on the last line and then add the new line, so the diff will show modifications on two lines. Same if you erase the last one, without this rule, you'll delete the last and modify the row before to delete the trailing comma.

I'm not a big fan of this rule either, but I have no problem to add it if someone finds it worth it. (Can be configured on VS to done automatically)

**Comment by pbarrios** [comment]
Thanks! 

Q: What about long lines of code? For example constructors with a lot parameters. I've seen that some developers use multiple lines while others just prefer do it in single long long line. What do you guys think is the best approach to take? I have not strong opinions about that.

File: CSharp/README.md (hunk 6 above)
Comment (AlphaGit) [inline CSharp/README.md]: No syntax errors, baby!

File: CSharp/README.md (hunk 9 above)
Comment (AlphaGit) [inline CSharp/README.md]: Same here, not a big fan, but I see the benefits. I was just asking for a general opinion... which seems to be against dangling comma.

**Comment by AlphaGit** [comment]
@pbarrios I believe long lines of code should be discouraged in general. However, it's really about complexity and missing stuff, not about the length of the line. In general, this is a symptom of a very complex line/class. I believe they should be broken down or refactored out.

I would be ok with making an exception for strings (pretty much like we did in the JavaScript guidelines).

File: CSharp/README.md (hunk 10 above)
Comment (AlphaGit) [inline CSharp/README.md]: I'll check what you did, I'm not sure I have a good way of explaining that. Basically, we want to avoid "weird-looking spacing".

Good:

```csharp
var result = (4 / 2) + 6;
var element = myArray[result];
```

Bad:

```csharp
var result =(4 / 2) + 6; // missing space to the left of opening parenthesis
var result = ( 4 / 2) + 6; // extra space to the right of opening parenthesis
var result = (4 / 2 ) + 6; // extra space to the left of closing parenthesis
var result = (4 / 2)+ 6; // missing space to the right of closing parenthesis

var element = myArray [result] + 1; // extra space to the left of opening bracket
var element = myArray[ result] + 1; // extra space to the right of opening bracket
...
```

You get the idea hehe, there are lots of combinations, and I don't know really how to word them all together.

Comment (mvazquez-ms) [inline CSharp/README.md]: To avoid writing things like ' Code must not contain space after new keyword in implicitly typed array allocation'

I think we should add a comprehensive piece of code that have examples about every spacing scenario and add a single rule saying:

_Spacing should comply with StyleCop rules SA1001 to SA1026_

File: CSharp/README.md (hunk 11 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: Good one, thanks!

File: CSharp/README.md (hunk 9 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: Ok let's keep that in mind to revisit later. We can add a favor or against style.

File: CSharp/README.md (hunk 2 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: I do not have a strong preference on this one too. In order to move forward and taking into account that we don't have a consensus here I will suggest anyone with strong arguments to create a ticket or PR with a removal or modification of this style. In that way we all be able to concentrate in one rule and take a decision. Thanks all for the feedback.

File: CSharp/README.md (hunk 10 above)
Comment (matiasbeckerle) [inline CSharp/README.md]: No problem with this one. I will add more examples. In the near future, if anyone has a nice idea for rephrasing the style, we can modify it.

**Comment by matiasbeckerle** [comment]
I think we might have covered everything discussed on the PR. Thank you all again, feel free to create new tickets or PRs if you want to address something particular. I hope to see you all next week, probably with naming styles!

Please: one approval at least? 😄

**Comment by gastonmancini** [comment]
Great work @matiasbeckerle !!!
